### PR TITLE
Registration

### PR DESCRIPTION
**Pull request by thepseudoartist**

### CODE REVIEW

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import retrofit2.Response;
+
+public class ManualSignUpActivity extends AppCompatActivity {
+
+    private EditText inputUserName, inputPhone, inputPassword, inputConfirm, inputFName, inputLName, inputEmail;
+    private TextInputLayout inputUserNameLayout, inputPhoneLayout, inputPasswordLayout, inputConfirmLayout, inputFNameLayout, inputLNameLayout, inputEmailLayout;
+    private Button signUpProceed, back;
+    private ProgressBar signUpProgressBar;
+    private LinearLayout layoutFirst, layoutSecond;
+    private Boolean proceed = true, first = true;
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Declare here,assigin values in initView() function.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        signUpProgressBar = findViewById(R.id.signUpProgressBar);
+
+        layoutFirst = findViewById(R.id.lay_first);
+        layoutSecond = findViewById(R.id.lay_second);
+
+        back.setEnabled(false);
+
+        signUpProceed.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: implement onClickListener in the class and override onClick method and do everything there.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                    back.setEnabled(true);
+                    first = false;
+                    inputPassword.requestFocus();
+
+                } else {
+                    if (isEmpty()) {
+                        if (validatePassword())
+                            apiCall();
+                    } else {
+                        Toast.makeText(ManualSignUpActivity.this, "Some of the required fields are empty.", Toast.LENGTH_LONG).show();
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: "Some of the required fields are empty." define this string in strings.xml and use the reference here.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                    }
+                }
+
+            }
+        });
+
+        back.setOnClickListener(new View.OnClickListener() {
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: must be done on overriden onClick method.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+    private void setData() {
+        userName = inputUserName.getText().toString().trim();
+        password = inputPassword.getText().toString().trim();
+        firstName = inputFName.getText().toString().trim();
+        lastName = inputLName.getText().toString().trim();
+        mobile = inputPhone.getText().toString().trim();
+        email = inputEmail.getText().toString().trim();
+
+        userDetails.setUsername(userName);
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: I have directly set the data in userDetails object. You have to first validate each field and then only data musr be set.
Following details are necessary : 
1. Mobile verification of 10 numbers , Numbers starting from 6,7,8,9 must be only taken and numbers such as 1111111111,2222222222 etc must not be taken.
2. Email verification must be done . You can refer to the code of previous app for this.
3. Password length of min 6 chars
4. Name must not take numbers as input.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        inputLName.addTextChangedListener(new CustomTextWatcher(inputLName, inputLNameLayout, AppConstants.LAST_NAME));
+        inputUserName.addTextChangedListener(new CustomTextWatcher(inputUserName, inputUserNameLayout, AppConstants.USER_NAME));
+        inputEmail.addTextChangedListener(new CustomTextWatcher(inputEmail, inputEmailLayout, AppConstants.EMAIL));
+        inputPassword.addTextChangedListener(new CustomTextWatcher(inputPassword, inputPasswordLayout, AppConstants.PASSWORD));
+        inputPhone.addTextChangedListener(new CustomTextWatcher(inputPhone, inputPhoneLayout, AppConstants.MOBILE_NO));
+    }
+
+    private boolean validatePassword() {
+        if (inputPassword.getText().toString().trim().equals(inputConfirm.getText().toString().trim())) {
+            inputConfirmLayout.setError("Passwords didn't match.");
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: "Passwords didn't match." define this in strings.xml

File: app/src/main/java/nitrr/ecell/e_cell/activities/RegisterMainActivity.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
+package nitrr.ecell.e_cell.activities;
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/activities/RegisterMainActivity.java:1]: Use initView() for find view by id kind of things. and do same for on click as stated above.

File: app/src/main/java/nitrr/ecell/e_cell/model/UserDetails.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
+package nitrr.ecell.e_cell.model;
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/model/UserDetails.java:1]: Remove username and make the model as discussed in the meeting

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    @Override
+    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
+    }
+
+    @Override
+    public void afterTextChanged(Editable editable) {
+        String text_content = text.getText().toString().trim();
+
+        if (text_content.equals("")) {
+            layout.setError("Empty fields not allowed.");
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: define in strings.xml

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        if (field_name.equals(AppConstants.FIRST_NAME)) {
+            Pattern pattern = Pattern.compile("[a-zA-Z]+");
+
+            if (pattern.matcher(text_content).matches()) {
+                layout.setErrorEnabled(false);
+                text.requestFocus();
+            } else {
+                layout.setError("First Name should not contain digits or empty field.");
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: define in strings.xml

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+
+        if (field_name.equals(AppConstants.LAST_NAME)) {
+            Pattern pattern = Pattern.compile("[a-zA-Z]+");
+
+            if (pattern.matcher(text_content).matches()) {
+                layout.setErrorEnabled(false);
+            } else {
+                layout.setError("Last Name should not contain digits or empty field.");
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: same here

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+                text.requestFocus();
+            }
+        }
+
+        if (field_name.equals(AppConstants.EMAIL)) {
+            if (!Patterns.EMAIL_ADDRESS.matcher(text_content).matches()) {
+                layout.setError("Invalid Email Address.");
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: here also

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+                text.requestFocus();
+
+            } else {
+                layout.setErrorEnabled(false);
+            }
+        }
+
+        if (field_name.equals(AppConstants.PASSWORD)) {
+            if (text_content.length() != 8) {
+                layout.setError("Password's length should at least be eight.");
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: here also

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+            } else {
+                layout.setErrorEnabled(false);
+            }
+        }
+
+        if (field_name.equals(AppConstants.MOBILE_NO)) {
+            Pattern pattern = Pattern.compile("[0-9]+");
+            if (!pattern.matcher(text_content).matches() || text_content.length() != 10) {
+                layout.setError("Enter a valid mobile number.");
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: here also

File: app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+                Bundle parameters = new Bundle();
+                parameters.putString("fields", "id, first_name, last_name, email");
+
+                request.setParameters(parameters);
+                request.executeAsync();
+            }
+
+            @Override
+            public void onCancel() {
+                Toast.makeText(activity, "Facebook SignUp Cancelled.", Toast.LENGTH_LONG).show();
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java]: define in strings.xml

File: app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        try {
+
+            String id = jsonObject.getString("id");
+
+            URL avatar_url;
+            avatar_url = new URL("https://graph.facebook.com/" + id + "/picture?type=large");
+
+            bundle.putString("id", id);
+
+            if (jsonObject.has("first_name"))
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java]: keys must be defined in AppConstants

File: app/src/main/java/nitrr/ecell/e_cell/utils/PrefUtils.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        });
+        dialog.show();
+    }
+
+    public Map<String, String> getFbUserInfo() {
+
+        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(activity);
+        Map<String, String> userInfo = new HashMap<>();
+
+        userInfo.put("uid", prefs.getString("uid", null));
```
Comment (thakersamveg608) [inline app/src/main/java/nitrr/ecell/e_cell/utils/PrefUtils.java:63]: keys in AppConstants

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class ManualSignUpActivity extends AppCompatActivity {
+
+    private EditText inputUserName, inputPhone, inputPassword, inputConfirm, inputFName, inputLName, inputEmail;
+    private TextInputLayout inputUserNameLayout, inputPhoneLayout, inputPasswordLayout, inputConfirmLayout, inputFNameLayout, inputLNameLayout, inputEmailLayout;
+    private Button signUpProceed, back;
+    private ProgressBar signUpProgressBar;
```
Comment (thepseudoartist) [inline app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Delete the Progress bar. :smile:

File: app/src/main/res/layout/manual_signup_layout.xml [hunk 19]
```diff
@@ -0,0 +1,325 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Vrihas123) [inline app/src/main/res/layout/manual_signup_layout.xml:1]: Do formatting of this xml

File: app/src/main/res/layout/manual_signup_layout.xml [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+            android:textSize="18sp" />
+
+
+        <TextView
+            android:id="@+id/sign_up_three"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginBottom="30dp"
+            android:layout_marginEnd="10dp"
+
```
Comment (Vrihas123) [inline app/src/main/res/layout/manual_signup_layout.xml]: remove this space

File: app/src/main/res/layout/manual_signup_layout.xml [hunk 21]
```diff
@@ -0,0 +1,325 @@
...
+                android:id="@+id/lay_first"
+                android:layout_width="350dp"
+                android:layout_height="wrap_content"
+                android:layout_alignParentTop="true"
+                android:layout_margin="30dp"
+                android:animateLayoutChanges="true"
+                android:background="@drawable/layout_background"
+                    android:orientation="vertical">
+
+                    <android.support.design.widget.TextInputLayout
```
Comment (Vrihas123) [inline app/src/main/res/layout/manual_signup_layout.xml]: combine similar attributes into a style and wherever required, use that style. Search for common attributes in all the xmls and put it in the styles. I am mentioning here the changes to be done in all the xmls, not particularly for this one.

File: app/src/main/res/layout/register_main.xml [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+        android:src="@mipmap/ecell_colored"
+        android:transitionName="imageTrans" />
+
+    <LinearLayout
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:orientation="vertical">
+
+        <Button
```
Comment (Vrihas123) [inline app/src/main/res/layout/register_main.xml:24]: For example, each button here contains attributes like focusable, clickable, font family etc. combine this into a style.

File: app/src/main/res/drawable/button_style_border.xml [hunk 23]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+
+    <solid android:color="@color/colorBackground" />
+
```
Comment (Vrihas123) [inline app/src/main/res/drawable/button_style_border.xml:7]: formatting required in these drawable resource files also

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java (hunk 1 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Done!

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java (hunk 2 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java (hunk 3 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java (hunk 4 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java (hunk 5 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java (hunk 6 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/activities/ManualSignUpActivity.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/activities/RegisterMainActivity.java (hunk 7 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/activities/RegisterMainActivity.java:1]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/model/UserDetails.java (hunk 8 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/model/UserDetails.java:1]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java (hunk 9 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java (hunk 10 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java (hunk 11 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java (hunk 12 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java (hunk 13 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: Done

File: app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java (hunk 14 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/CustomTextWatcher.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java (hunk 15 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java]: Done

File: app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java (hunk 16 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/FacebookSignUp.java]: Done.

File: app/src/main/java/nitrr/ecell/e_cell/utils/PrefUtils.java (hunk 17 above)
Comment (thepseudoartist) [reply app/src/main/java/nitrr/ecell/e_cell/utils/PrefUtils.java:63]: Done.

File: app/src/main/res/layout/manual_signup_layout.xml (hunk 20 above)
Comment (thepseudoartist) [reply app/src/main/res/layout/manual_signup_layout.xml]: Done.

File: app/src/main/res/drawable/button_style_border.xml (hunk 23 above)
Comment (thepseudoartist) [reply app/src/main/res/drawable/button_style_border.xml:7]: What kind of formatting, can you be more specific sir?

**Review by thakersamveg608 — COMMENTED** [review]
PR passed.

**Pull request closed without merging**